File: python/cut_off_trees_golf_event_675.py
# BFS, Map, overtime
class Solution:
    def cutOffTree(self, forest):
        """
        :type forest: List[List[int]]
        :rtype: int
        """
        height = len(forest)
        if height == 0:
            return 0
        width = len(forest[0])
        if width == 0:
            return 0
        trees = []
        for i in range(height):
            for j in range(width):
                if forest[i][j] > 0:
                    trees.append((forest[i][j], i, j))
        trees.sort(key=lambda x: x[0])
        # print(trees)
        count = 0
        sr = 0
        sc = 0
        for (_, tr, tc) in trees:
            dist = self.dist(forest, sr, sc, tr, tc)
            if dist < 0:
                return -1
            count += dist
            sr, sc = tr, tc
        return count

    def dist(self, forest, sr, sc, tr, tc):
        # print(sr, sc, tr, tc)
        height = len(forest)
        width = len(forest[0])
        queue = collections.deque()
        queue.append((sr, sc, 0))
        seen = [[False for i in range(width)] for j in range(height)]
        seen[sr][sc] = True
        # print(seen)
        # print(queue)
        while queue:
            (nowr, nowc, nowdist) = queue.popleft()
            # print(nowr, nowc, nowdist)
            if nowr == tr and nowc == tc:
                return nowdist
            for nr, nc in ((nowr - 1, nowc), (nowr + 1, nowc), (nowr, nowc - 1), (nowr, nowc + 1)):
                if (0 <= nr < height and 0 <= nc < width and seen[nr][nc] == False and forest[nr][nc]):
                    seen[nowr][nowc] = True
                    queue.append((nr, nc, nowdist + 1))
            # print(queue)
        return -1

# My method, 8s


class Solution:
    def cutOffTree(self, forest):
        """
        :type forest: List[List[int]]
        :rtype: int
        """
        height = len(forest)
        if height == 0:
            return 0
        width = len(forest[0])
        if width == 0:
            return 0
        trees = []
        for i in range(height):
            for j in range(width):
                if forest[i][j] > 0:
                    trees.append((forest[i][j], i, j))
        trees.sort(key=lambda x: x[0])
        # print(trees)
        count = 0
        sr = 0
        sc = 0
        for (_, tr, tc) in trees:
            dist = self.dist(forest, sr, sc, tr, tc)
            if dist < 0:
                return -1
            count += dist
            sr, sc = tr, tc
        return count

    def dist(self, forest, sr, sc, tr, tc):
        # print(sr, sc, tr, tc)
        height = len(forest)
        width = len(forest[0])
        queue = collections.deque()
        queue.append((sr, sc, 0))
        seen = [[False for i in range(width)] for j in range(height)]
        # seen[sr][sc] = True
        # print(seen)
        # print(queue)
        while len(queue) != 0:
            (nowr, nowc, nowdist) = queue.popleft()
            # print(nowr, nowc, nowdist)
            if nowr < 0 or nowr >= height or nowc < 0 or nowc >= width or seen[nowr][nowc] or forest[nowr][nowc] <= 0:
                continue
            if nowr == tr and nowc == tc:
                return nowdist
            seen[nowr][nowc] = True
            queue.extend([(nowr + i, nowc + j, nowdist + 1)
                          for i, j in zip([1, -1, 0, 0], [0, 0, 1, -1])])
            # print(queue)
        return -1

File: amazon/java/cut_off_trees_golf_event_675.java
// BFS, map, time O((r*c)^2) space O(r*c)
class Solution {
    public int cutOffTree(List<List<Integer>> forest) {
        List<int[]> trees = new ArrayList();
        for (int r = 0; r < forest.size(); r++) {
            for (int c = 0; c < forest.get(r).size(); c++) {
                int value = forest.get(r).get(c);
                if (value > 1) {
                    trees.add(new int[] {value, r, c});
                }
            }
        }
        Collections.sort(trees, (a, b) -> Integer.compare(a[0], b[0]));

        int ans = 0, sr = 0, sc = 0;
        for (int[] tree: trees) {
            int dis = dist(forest, sr, sc, tree[1], tree[2]);
            if (dis < 0) return -1;
            ans += dis;
            sr = tree[1];
            sc = tree[2];

        }
        return ans;
    }

    public int dist(List<List<Integer>> forest, int sr, int sc, int tr, int tc) {
        int[] dr = {-1, 1, 0, 0};
        int[] dc = {0, 0, -1, 1};
        int height = forest.size();
        int width = forest.get(0).size();
        Queue<int[]> queue = new LinkedList();
        // Queue includes the array of row#, column#, and distance to source
        queue.offer(new int[] {sr, sc, 0});
        boolean[][] seen = new boolean[height][width];
        seen[sr][sc] = true;
        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            if (cur[0] == tr && cur[1] == tc) return cur[2];
            for (int di = 0; di < 4; di++) {
                int r = cur[0] + dr[di];
                int c = cur[1] + dc[di];
                if (0 <= r && r < height && 0 <= c && c < width && !seen[r][c] && forest.get(r).get(c) > 0) {
                    seen[r][c] = true;
                    queue.offer(new int[] {r, c, cur[2]+1});
                }
            }
        }
        return -1;
    }
}